[MID-47] add event page

Yivgha opened the pull request:
Added event page styles for all devices, data fetch, prisma and seeds logic, infinite scroll on events when more than 6 items, implemented EventCard and EventPopUp draft from other PRs. All data is filtered and sorted by date, showing only active events. Also added NoInfoToShow reusable component and filter buttons with conditional icons.

[Trello [MID-47]](https://trello.com/c/cPnzyL0C/47-%D1%8E%D0%B7%D0%B5%D1%80-%D1%81%D1%82%D0%BE%D1%80%D1%96%D0%BD%D0%BA%D0%B0-%D0%B7-%D1%83%D1%81%D1%96%D0%BC%D0%B0-%D0%BF%D0%BE%D0%B4%D1%96%D1%8F%D0%BC%D0%B8)

Yivgha committed:
[MID-47] add page layout and basic components

Yivgha committed:
Merge branch 'dev' of https://github.com/keenethics/mayak into MID-47_add-events-page

Yivgha committed:
[MID-47] add filter by date

Yivgha committed:
[MID-47] add icon to filter button

Yivgha committed:
[MID-47] add filter data by month

Yivgha committed:
Merge branch 'dev' of https://github.com/keenethics/mayak into MID-47_add-events-page

Yivgha committed:
[MID-47] merge dev

Yivgha committed:
[MID-47] fix seed error

Yivgha committed:
[MID-47] add EventCard to page

Yivgha committed:
[MID-47] add month param to page url

+little refactor to NoInfoToShow component

Yivgha committed:
[MID-47] fix styles for filter button

Yivgha committed:
Merge branch 'dev' of https://github.com/keenethics/mayak into MID-47_add-events-page

Yivgha committed:
[MID-47] merge dev and run lint fix

Yivgha committed:
Merge branch 'dev' of https://github.com/keenethics/mayak into MID-47_add-events-page

Yivgha committed:
[MID-47] refactor filters and fetching on events page

Yivgha committed:
[MID-47] small fixes

Yivgha committed:
Merge branch 'dev' of https://github.com/keenethics/mayak into MID-47_add-events-page

Yivgha committed:
[MID-47] fix icons import after merge dev

Yivgha committed:
[MID-47] run lint fix

Yivgha committed:
[MID-47] fix urlsearchparam

Yivgha committed:
[MID-47] lint fix

Yivgha committed:
[MID-47] fix data on first render

Yivgha committed:
[MID-47] fix styles and refactor

Yivgha committed:
[MID-47] small changes

Yivgha committed:
Merge branch 'dev' of https://github.com/keenethics/mayak into MID-47_add-events-page

Yivgha committed:
[MID-47] small style fix

Dimirolz commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+  const [activeMonthNumber, setActiveMonthNumber] = useState('');
+  const [hoveredIndexes, setHoveredIndexes] = useState(new Array(months.length));
+
+  const currentMonth = new Date().getMonth() + 1;
```
@Yivgha Let's move static data outside of the component

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+  }
+
+  // Get 6 basic month names for filters, starting from current month
+  useEffect(() => {
+    const getNextMonths = () => {
+      const nextMonths = [];
+      for (let i = 0; i <= monthsAhead; i += 1) {
+        const nextMonthIndex = (currentMonth + i - 1) % 12;
+        const nextMonthName = monthNames[nextMonthIndex];
+        setActiveMonthNumber(monthNames[i] === currentMonth ? monthNames[i] : '');
+        nextMonths.push(nextMonthName);
+      }
+      return nextMonths;
+    };
+    const nextMonths = getNextMonths();
+    setMonths(nextMonths);
+    // eslint-disable-next-line
+  }, [currentMonth]);
```
@Yivgha there is multiple things 
1. I don't see a reason those logic to be an effect. It can be calculated statically outside of the component
2. logic is too complicated for such case and it leads for mistakes. monthNames[i] is object, when currentMonth is a number, so comparing them with === is unusfull. Let's rewrite this function with .slice method of arrays. (you can check that monthsAhead is out of the length of the array and make concatination of 2 slices).
3. We should rid all of the states and make them static outside of the component, they are not depend on any reactive things

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+  const [activeIndex, setActiveIndex] = useState(0);
+  const [activeMonth, setActiveMonth] = useState(months[0]);
+  const [activeMonthNumber, setActiveMonthNumber] = useState('');
+  const [hoveredIndexes, setHoveredIndexes] = useState(new Array(months.length));
```
@Yivgha we should remove it in order to make this logic via css

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+      return <CheckMark key={`checkmark+${index}`} className="h-4 w-4 transition-all" />;
+    }
+    if (hoveredIndexes[index] && activeIndex !== index && (semiorange.hover || semiorange.focused)) {
+      return <Search key={`searchicon+${index}`} className="h-4 w-4 transition-all " />;
```
@Yivgha Let's always render Search icon for buttons that are not === activeIndex, but hide it with `hidden` by default. Then we can make it `group-hover:block` when user hover over button.

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+              className={cn(activeIndex === month.index && activeButtonStyles, 'group w-fit')}
+              key={month.index}
+              activemonth={activeMonth}
+              onMouseEnter={() => handleMouseEnter(month.index)}
+              onMouseLeave={() => handleMouseLeave(month.index)}
+              onFocus={() => {
+                handleMouseEnter(month.index);
+              }}
+              onBlur={() => handleMouseLeave(month.index)}
+              onMouseDown={() => handleMouseEnter(month.index)}
+              onMouseUp={() => handleMouseLeave(month.index)}
```
@Yivgha we should avoid all of this event handlers using native hover

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/config.js:
```diff
@@ -0,0 +1,15 @@
+export const monthsAhead = 5;
+export const monthNames = [
+  { index: 0, name: 'січень', en: 'january' },
+  { index: 1, name: 'лютий', en: 'february' },
+  { index: 2, name: 'березень', en: 'march' },
+  { index: 3, name: 'квітень', en: 'april' },
+  { index: 4, name: 'травень', en: 'may' },
+  { index: 5, name: 'червень', en: 'june' },
+  { index: 6, name: 'липень', en: 'july' },
+  { index: 7, name: 'серпень', en: 'august' },
+  { index: 8, name: 'вересень', en: 'september' },
+  { index: 9, name: 'жовтень', en: 'october' },
+  { index: 10, name: 'листопад', en: 'november' },
+  { index: 11, name: 'грудень', en: 'december' },
+];
```
@Yivgha it's up to you, to create separate file, but if it used only in Event component, I would add it in the component file for the locality.

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+import { monthsAhead, monthNames } from './config';
+
+export function EventFilter() {
+  const [months, setMonths] = useState([]);
```
@Yivgha we can move it outside of the component, it shouldn't be reactive

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+
+export function EventFilter() {
+  const [months, setMonths] = useState([]);
+  const [activeIndex, setActiveIndex] = useState(0);
+  const [activeMonth, setActiveMonth] = useState(months[0]);
+  const [activeMonthNumber, setActiveMonthNumber] = useState('');
```
@Yivgha for me it looks like those 3 state are all the same. we can use only one state and others would be derived from it

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+      setActiveMonthNumber(monthFromQuery);
+      setActiveIndex(monthFromQuery - 1);
+      setActiveMonth(months.filter(item => item.index === parseInt(monthFromQuery, 10)));
+      allEvents({ month: monthFromQuery, take: '', lastCursor: '' });
```
@Yivgha we should avoid calling those api directly when we already set up React Query.

When you will change queryKey reactive variable, then RQ will automatically refetch the data

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+      setActiveIndex(currentMonth - 1);
+      deleteParam();
+      addParam(currentMonth.toString());
+      allEvents({ month: currentMonth, take: '', lastCursor: '' });
```
@Yivgha same here for calling api, we shouldn't

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+    setActiveMonthNumber(calcMonth);
+    deleteParam();
+    addParam(calcMonth);
+    allEvents({ month: calcMonth, take: '', lastCursor: '' });
```
@Yivgha  same here for calling api, we shouldn't

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+    allEvents({ month: calcMonth, take: '', lastCursor: '' });
+  };
+
+  const handleMouseEnter = index => {
+    const newHoveredIndexes = [...hoveredIndexes];
+    newHoveredIndexes[index] = true;
+    setHoveredIndexes(newHoveredIndexes);
+  };
+
+  const handleMouseLeave = index => {
+    const newHoveredIndexes = [...hoveredIndexes];
+    newHoveredIndexes[index] = false;
+    setHoveredIndexes(newHoveredIndexes);
+  };
```
@Yivgha let's implement it via css hover

Yivgha replied:
changed

alex-popov-tech commented on src/app/(app)/event/page.js:
```diff
@@ -0,0 +1,28 @@
+        <Caption type="c2" className="text-p4 font-bold uppercase text-primary-600">
+          найближчі місяці
+        </Caption>
+        <EventFilter />
```
🟠 maybe 'Events' or 'EventsSection' is better name for this component? because its not only filter, its filter + cards

Yivgha replied:
changed

alex-popov-tech commented on src/app/_components/Event/Card.js:
```diff
@@ -1,11 +1,13 @@
 import { parseDate } from '@/utils/parseDate';
 import { Label } from '../Label';
 import { OverflownText } from '../OverflownText';
+import { EventLinkModal } from '../EventLinkModal';
```
🟠 as per conventions all imports should be aliased

Yivgha replied:
changed

alex-popov-tech commented on src/app/_components/Event/Card.js:
```diff
@@ -35,11 +39,13 @@ function transformData(event) {
-export default function EventCard({ event }) {
-  const { title, organizerName, tags, priceText, locationText, date, time, locationLink } = transformData(event);
+export function EventCard({ event }) {
+  const [isFeedbackOpen, setFeedbackOpen] = useState(false);
```
🟠 isModalOpen

Yivgha replied:
changed

alex-popov-tech commented on src/app/_components/Event/Card.js:
```diff
@@ -48,12 +54,16 @@ export default function EventCard({ event }) {
   );
 
-  const tagsElements = tags.map(tag => (
+  function toggleFeedback() {
```
🟠 toggleModal

Yivgha replied:
changed

alex-popov-tech commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+import { allEvents, useEventSetParam } from '@hooks';
+import { cn } from '@/utils/cn';
+import { PillButton } from '../PillButton';
+import { buttonColorVariant } from '../PillButton/style';
```
🟠  please pull latest dev

Yivgha replied:
changed

alex-popov-tech commented on src/app/_components/Event/EventFilter.js:
```diff
@@ -0,0 +1,179 @@
+import { cn } from '@/utils/cn';
+import { PillButton } from '../PillButton';
+import { buttonColorVariant } from '../PillButton/style';
+import { NoInfoToShow } from '../NoInfoToShow';
```
🟠 please use import aliases due to conventions

Yivgha replied:
changed

alex-popov-tech commented on src/app/_hooks/api/useEventList.js:
```diff
@@ -0,0 +1,6 @@
+import ky from 'ky';
+
+export const allEvents = async ({ month, take, lastCursor }) => {
```
🟠 is it used anywhere?

Yivgha replied:
When fetching data on event page

alex-popov-tech commented on src/app/_hooks/api/useEventSetParam.js:
```diff
@@ -0,0 +1,25 @@
+  const newParams = new URLSearchParams(searchParams);
+
+  const addParam = value => {
+    if (searchParams.get(param) && param.month === 'month') {
```
🟠 could existing hook for params be used in this case?

Yivgha replied:
replaced

alex-popov-tech commented on src/app/api/event/route.js:
```diff
@@ -1,6 +1,80 @@
+
+    return new Response(JSON.stringify(data), { status: 200 });
+  } catch (error) {
+    return new Response(JSON.stringify({ error: error.message }), { status: 403 });
```
🟠 why 403? 🤔

Yivgha replied:
changed to 404

Yivgha committed:
[MID-47] refactor after review

Yivgha committed:
Merge branch 'dev' of https://github.com/keenethics/mayak into MID-47_add-events-page

Yivgha committed:
[MID-47] change imports and hooks

Dimirolz commented on src/app/_components/Event/EventSection.js:
```diff
@@ -0,0 +1,154 @@
+
+const calcPureMonthIndex = parseInt(currentMonth - 1, 10);
+const firstMonth = monthNames.slice(calcPureMonthIndex, calcPureMonthIndex + 1);
+const restMonth = monthNames.slice(calcPureMonthIndex + 1, calcPureMonthIndex + 6);
```
@Yivgha you should handle case when calcPureMonthIndex + 6 is bigger than monthNames. Imagine that today is December, which months you want to have here? (hint: little modular arithmetic)

Yivgha replied:
changed

Dimirolz replied:
@Yivgha it's not handled yet. calcPureMonthIndex + 6 can go beyond length of the monthNames, you should handle that.

Yivgha replied:
changed

Dimirolz commented on src/app/api/event/route.js:
```diff
@@ -1,6 +1,80 @@
+    const lastCursor = url.searchParams.get('lastCursor');
+    const queryMonth = url.searchParams.get('month');
+
+    const filteredQueryMonth = parseInt(queryMonth < 10 ? `0${queryMonth}` : queryMonth, 10);
```
@Yivgha is there any sense for this ternaries operator? You parsing string immediately after padding it with zero

Dimirolz replied:
Same for default radix, we can safely remove it

Yivgha replied:
> @Yivgha is there any sense for this ternaries operator? You parsing string immediately after padding it with zero

yeah, because from query we get number without 0 and to pass month to new Date we should have it. if month is 10+ we don't need it.

Dimirolz replied:
@Yivgha sounds reasonable, but this code you are doing opposite thing.

Imagine that you have queryMonth = 4.
1. parseInt(4 < 10 ? `0${4}` : 4, 10);
2. 4 is smaller than 10, so we have parseInt( `04`, 10);
3. When parseInt parses string '04' we will get number 4.

So in this line we are not doing anything useful. It's the same if we just wrote `parseInt(queryMonth, 10)`.

p.s. btw, you are passing after that number without 0 and everything works

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventSection.js:
```diff
@@ -0,0 +1,154 @@
+              }}
+              icon={selectButtonIcon(month.index)}
+            >
+              {month.name.charAt(0).toUpperCase() + month.name.slice(1)}
```
@Yivgha Do you have capitalize util in project? If not, it can be useful to create one

Yivgha replied:
changed

Dimirolz commented on src/app/_components/NoInfoToShow.js:
```diff
@@ -0,0 +1,15 @@
+export function NoInfoToShow({ text }) {
+  return (
+    <div className="no-wrap lg:w-max-[900px] flex h-[250px] w-full flex-shrink-0 flex-col items-center justify-center gap-4 bg-gray-100 md:h-[300px] md:gap-10 lg:h-[400px] lg:gap-10">
+      <SadIcon className="h-[77px] w-[77px] text-gray-300 md:h-[121px] md:w-[121px] lg:h-[121px] lg:w-[121px]" />
```
@Yivgha There is style duplication for md and lg breakpoints

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventSection.js:
```diff
@@ -0,0 +1,154 @@
+const filteredMonths = firstMonth.concat(restMonth);
+
+export function EventSection() {
+  const [activeMonth, setActiveMonth] = useState(parseInt(filteredMonths[0].index + 1, 10));
```
@Yivgha This is up to you, but I would suggest to remove radix 10, it’s default for parseInt

Yivgha replied:
we have eslint rule, which demands having radix

Dimirolz commented on src/app/api/event/route.js:
```diff
@@ -1,6 +1,80 @@
+          lte: filteredQueryMonth === today.getMonth() + 1 ? endOfMonth : endOfNextMonth,
+        },
+      },
+      take: take ? parseInt(take, 10) : 6,
```
@Yivgha default radix

Yivgha replied:
we have eslint rule, which demands having radix

Dimirolz commented on src/app/api/event/route.js:
```diff
@@ -1,6 +1,80 @@
+    const cursor = lastPostInResults.id;
+
+    const nextPage = await prisma.event.findMany({
+      take: take ? parseInt(take, 10) : 6,
```
@Yivgha default radix

Yivgha replied:
we have eslint rule, which demands having radix

Dimirolz commented on src/app/_components/Event/EventSection.js:
```diff
@@ -0,0 +1,154 @@
+        </ul>
+      </div>
+
+      {(isLoading || isFetchingNextPage) && <CircularProgress />}
```
@Yivgha can we use skeleton here? Page is jumping every time when I’m changing month in filter for the first time.

Yivgha replied:
created Skeletoncard

Dimirolz replied:
@Yivgha I guess we don't need Progress in the same time with Skeleton then.

Yivgha replied:
deleted Progress

Dimirolz commented on src/app/_components/Event/EventSection.js:
```diff
@@ -0,0 +1,154 @@
+  { index: 11, name: 'грудень', en: 'december' },
+];
+
+const calcPureMonthIndex = parseInt(currentMonth - 1, 10);
```
@Yivgha you can safely remove radix

Yivgha replied:
we have eslint rule, which demands having radix

Dimirolz commented on src/app/_components/Event/EventSection.js:
```diff
@@ -0,0 +1,154 @@
+  }, [hasNextPage, inView]);
+
+  // Filter data based on selected month
+  const handleFilter = newMonth => {
```
@Yivgha wrap this one into useCallaback to preserve function identity on rerenders

Yivgha replied:
done

Dimirolz commented on src/app/_components/Event/EventSection.js:
```diff
@@ -0,0 +1,154 @@
+  const [activeMonth, setActiveMonth] = useState(parseInt(filteredMonths[0].index + 1, 10));
+
+  const searchParams = useSearchParams();
+  const { add: addParam, remove: removeParam } = useSetParam('month');
```
@Yivgha we don't need two of them, we can just use `replace` from useSetParam hook

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventSection.js:
```diff
@@ -0,0 +1,154 @@
+  });
+
+  // Get data on first render or on url open
+  useEffect(() => {
+    const monthFromQuery = searchParams.get('month');
+
+    if (monthFromQuery && monthFromQuery !== currentMonth) {
+      // If the month is defined in the query and is not equal to the current month
+      setActiveMonth(monthFromQuery);
+      removeParam();
+      addParam(monthFromQuery);
+    } else if (!monthFromQuery || monthFromQuery === undefined) {
+      setActiveMonth(currentMonth);
+      removeParam();
+      addParam(currentMonth);
+    }
+    // eslint-disable-next-line
+  }, [currentMonth]);
```
@Yivgha this useEffect is redundant and creates flickering. Let's move this logic into the closure that will set default value for `activeMonth`

Like that 

```js
 const searchParams = useSearchParams();
  const [activeMonth, setActiveMonth] = useState(() => {
  const monthFromQuery = searchParams.get('month');
    return monthFromQuery || filteredMonths[0].index + 1;
  });

```

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventSection.js:
```diff
@@ -0,0 +1,154 @@
+    'pointer-events-none border-secondary-300 bg-secondary-300 font-semibold text-gray-900': semiorange.active,
+  });
+
+  function selectButtonIcon(index) {
+    if (activeMonth - 1 === index) {
+      return (
+        <CheckMark
+          key={`checkmark+${index}`}
+          className={cn(`block h-4 w-4 transition-all group-hover:hidden group-focus:hidden`)}
+        />
+      );
+    }
+
+    return (
+      <Search
+        key={`searchicon+${index}`}
+        className="hidden h-4 w-4 transition-all group-hover:block group-focus:block"
+      />
+    );
+  }
```
@Yivgha I would inline this logic into PillButton, otherwise you should wrap it with useCallback

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventSection.js:
```diff
@@ -0,0 +1,154 @@
+
+  const { ref, inView } = useInView();
+
+  const { semiorange } = buttonColorVariant.eventFilter;
+
+  const activeButtonStyles = cn({
+    'pointer-events-none border-secondary-300 bg-secondary-300 font-semibold text-gray-900': semiorange.active,
+  });
```
@Yivgha move static outside of the component, to make rerenders lighter

Yivgha replied:
changed

Dimirolz commented on src/app/_components/Event/EventSection.js:
```diff
@@ -0,0 +1,154 @@
+  // Filter data based on selected month
+  const handleFilter = newMonth => {
+    setActiveMonth(newMonth);
+    removeParam();
+    addParam(newMonth.toString());
```
@Yivgha change to `replace` function

Yivgha replied:
changed

Yivgha committed:
[MID-47] changes after review

Yivgha committed:
[MID-47] add SkeletonCard and little refactoring

Yivgha committed:
[MID-47] small changes

Dimirolz commented on src/app/_components/Event/EventSection.js:
```diff
@@ -0,0 +1,142 @@
+const restMonth = monthNames.slice(calcPureMonthIndex + 1, calcPureMonthIndex + 6);
+const filteredMonths = firstMonth.concat(restMonth);
+
+const fakeArray = [1, 2, 3, 4, 5, 6];
```
@Yivgha it's not something that people often do for such cases. Previous iteration with `new Array(6).fill(0).map` was more idiomatic

alex-popov-tech replied:
or `_.range`

Yivgha replied:
with new Array(6).fill(0).map client throws errors with key of each elements: it dublicates and none of hacks like Math.floor(Math.random) don't work.

Dimirolz replied:
@Yivgha you should use array indexes and not values in this case

Dimirolz commented on src/app/_components/Event/SkeletonCard.js:
```diff
@@ -0,0 +1,10 @@
+const fakeCardArray = [7, 8, 9, 10, 11, 12, 13, 14];
```
@Yivgha there is no need for separate array with this numbers. `new Array` is more idiomatic for iterating

alex-popov-tech replied:
or `_.range`

Yivgha replied:
same as above

Dimirolz commented on src/utils/common.js:
```diff
@@ -32,3 +32,5 @@ export const formatPhoneNumber = phoneNumber => {
   return `${countryCode} (${areaCode}) ${firstPart} ${secondPart}`;
 };
+
+export const capitalizeFirstLetter = inputString => inputString.charAt(0).toUpperCase() + inputString.slice(1);
```
@Yivgha let's rename it to just `capitalize` it's more convenient, many utility libs has this name for such function

alex-popov-tech commented on src/app/api/event/route.js:
```diff
@@ -1,6 +1,81 @@
+
+    return new Response(JSON.stringify(data), { status: 200 });
+  } catch (error) {
+    return new Response(JSON.stringify({ error: error.message }), { status: 404 });
```
can you please make an example what can i send to this route to recieve 404? it looks like a 'list' type of thing, which is not supposed to return 404 🤔 

also wrapping so much with this try-catch you could mask potential server errors with 404, maybe default handler with 500 will suite better here?

Yivgha replied:
added withErrorHandler

Yivgha committed:
[MID-47] add error handler and month func fix

Yivgha committed:
Merge branch 'dev' of https://github.com/keenethics/mayak into MID-47_add-events-page

Yivgha committed:
[MID-47] fix after TherapyPrices merge

Yivgha committed:
[MID-47] small fixes after atomicity testing

fix styles in EventCard while no tag
link in EventModal isn't required
in AP can add event and show in list/site

Yivgha committed:
Merge branch 'dev' of https://github.com/keenethics/mayak into MID-47_add-events-page
